### PR TITLE
BUG: Fix wrong SparseBlock initialization

### LINKED ISSUE #17198: AttributeError: 'IntBlock'/'FloatBlock'/etc. object has no attribute 'sp_index'
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
# AttributeError: 'IntBlock' object has no attribute 'sp_index'
s = pd.SparseSeries([1, 2])
print(s.where(s >= 2, 0))
# AttributeError: 'FloatBlock' object has no attribute 'sp_index'
df = pd.SparseDataFrame([[1, 2], [3, 4]])
print(df.quantile())
```
#### Problem description

On the surface, SparseSeries and SparseDataFrame should produce the same results as Series and DataFrame, respectively.

#### Expected Output

0    0
1    2
dtype: int64
0    2.0
1    3.0
Name: 0.5, dtype: float64

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.1.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 94 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: en
LOCALE: None.None

pandas: 0.20.1
pytest: 3.0.7
pip: 9.0.1
setuptools: 36.0.1
Cython: 0.25.2
numpy: 1.12.1
scipy: 0.19.0
xarray: None
IPython: 5.3.0
sphinx: 1.6.3
patsy: 0.4.1
dateutil: 2.6.1
pytz: 2017.2
blosc: None
bottleneck: 1.2.1
tables: 3.2.2
numexpr: 2.6.2
feather: None
matplotlib: 2.0.2
openpyxl: 2.4.7
xlrd: 1.0.0
xlwt: 1.2.0
xlsxwriter: 0.9.6
lxml: 3.7.3
bs4: 4.6.0
html5lib: 0.999
sqlalchemy: 1.1.9
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by Licht-T**
- [x] closes #17198
- [x] tests added / passed
    
    Passed the same tests which the current master branch pass.
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by jreback** [comment]
needs tests

**Review by jreback — CHANGES_REQUESTED** [review]
w/o tests I don't even know what you are trying to do. Further we don't use if/then like this at all, instead dispatch per-block. but pls add tests first.

**Comment by Licht-T** [comment]
@jreback Thanks for your comment. I've jast added test codes.

File: pandas/core/internals.py [hunk 1]
```diff
@@ -527,7 +535,7 @@ def f(m, v, i):
 
         return self.split_and_operate(None, f, False)
 
-    def astype(self, dtype, copy=False, errors='raise', values=None, **kwargs):
+    def astype(self, dtype, copy=True, errors='raise', values=None, **kwargs):
```
Comment (jreback) [inline pandas/core/internals.py]: pls don't change things like this, which completely changes semantics

File: pandas/core/internals.py [hunk 2]
```diff
@@ -442,12 +443,19 @@ def make_a_block(nv, ref_loc):
                     nv = _block_shape(nv, ndim=self.ndim)
                 except (AttributeError, NotImplementedError):
                     pass
-                block = self.make_block(values=nv,
-                                        placement=ref_loc, fastpath=True)
+
```
Comment (jreback) [inline pandas/core/internals.py]: I don't like this pattern at all, nv *should* be a sparse array at this point; the inference should just work. is that not the case?

File: pandas/core/internals.py [hunk 3]
```diff
@@ -1562,6 +1581,10 @@ def _nanpercentile(values, q, axis, **kw):
         result = self._try_coerce_result(result)
         if is_scalar(result):
             return ax, self.make_block_scalar(result)
+
```
Comment (jreback) [inline pandas/core/internals.py]: again this should be a sparse array, if its not then you need to implement this on the SparseBlock itself (and you can certainly call the super method). I don't like type checking *inside* a block.

File: pandas/core/internals.py [hunk 4]
```diff
@@ -2653,7 +2676,7 @@ def sp_index(self):
     def kind(self):
         return self.values.kind
 
-    def _astype(self, dtype, copy=False, raise_on_error=True, values=None,
+    def _astype(self, dtype, copy=True, raise_on_error=True, values=None,
```
Comment (jreback) [inline pandas/core/internals.py]: do not do this

File: pandas/core/sparse/frame.py [hunk 5]
```diff
@@ -321,8 +321,8 @@ def _apply_columns(self, func):
             data=new_data, index=self.index, columns=self.columns,
             default_fill_value=self.default_fill_value).__finalize__(self)
 
-    def astype(self, dtype):
-        return self._apply_columns(lambda x: x.astype(dtype))
+    def astype(self, dtype, copy=True, errors='raise'):
```
Comment (jreback) [inline pandas/core/sparse/frame.py]: does this match the super signature?

File: pandas/tests/sparse/test_frame.py [hunk 6]
```diff
@@ -1385,3 +1385,62 @@ def test_numpy_func_call(self):
                  'std', 'min', 'max']
         for func in funcs:
             getattr(np, func)(self.frame)
+
+    def test_where(self):
+        data = [[1, 2], [3, 4]]
```
Comment (jreback) [inline pandas/tests/sparse/test_frame.py]: add the issue number as a comment

File: pandas/tests/sparse/test_frame.py [hunk 7]
```diff
@@ -1385,3 +1385,62 @@ def test_numpy_func_call(self):
...
+
+        sparse_df = SparseDataFrame(data)
+        result = sparse_df.where(sparse_df >= 2)
+
+        dense_df = DataFrame(data)
```
Comment (jreback) [inline pandas/tests/sparse/test_frame.py]: so these return *dense* frames????

File: pandas/tests/sparse/test_frame.py [hunk 8]
```diff
@@ -1385,3 +1385,62 @@ def test_numpy_func_call(self):
...
+        expected = dense_df.where(dense_df >= 2, other)
+
+        tm.assert_frame_equal(result, expected)
+
+    def test_quantile(self):
+        data = [[1, 1], [2, 10], [3, 100], [4, 100]]
+        q = 0.1
+
+        sparse_df = SparseDataFrame(data)
+        result = sparse_df.quantile(q)
```
Comment (jreback) [inline pandas/tests/sparse/test_frame.py]: these are very odd tests, why are you returning dense frames?

File: pandas/tests/sparse/test_series.py [hunk 9]
```diff
@@ -1384,6 +1384,41 @@ def test_numpy_func_call(self):
...
                 getattr(np, func)(getattr(self, series))
 
+    def test_where(self):
+        data = [1, 2, 3, 4]
+
+        sparse = SparseSeries(data)
+        result = sparse.where(sparse >= 2)
+
+        dense = Series(data)
+        expected = dense.where(dense >= 2)
```
Comment (jreback) [inline pandas/tests/sparse/test_series.py]: same questions as above

**Review by jreback — CHANGES_REQUESTED** [review]
first need to get the tests correct

**Comment by Licht-T** [comment]
@jreback Thanks for your review! I've found the initial solution is not enough through fixing tests.
Also I've modified the solution!

File: pandas/core/internals.py [hunk 10]
```diff
@@ -978,7 +984,7 @@ def f(m, v, i):
 
         return [self.make_block(new_values, fastpath=True)]
 
-    def coerce_to_target_dtype(self, other):
+    def coerce_to_target_dtype(self, other, copy=False):
```
Comment (jreback) [inline pandas/core/internals.py]: hmm, I dont think this actually helps, why did you add it?

File: pandas/core/internals.py [hunk 11]
```diff
@@ -2609,10 +2634,62 @@ class SparseBlock(NonConsolidatableMixIn, Block):
...
     _ftype = 'sparse'
     _holder = SparseArray
 
+    def __init__(self, values, placement, ndim=None, fastpath=False, **kwargs):
+        super(SparseBlock, self).__init__(values, placement,
+                                          ndim, fastpath,
+                                          **kwargs)
+
+        dtype = self.values.sp_values.dtype
+
```
Comment (jreback) [inline pandas/core/internals.py]: this is pretty complex, pls simplify

**Review by jreback — CHANGES_REQUESTED** [review]
you are changing lots of things. can you do this step by step rather that all at once. you are touching some pretty gnarly code.

File: pandas/core/internals.py (hunk 10 above)
Comment (Licht-T) [reply pandas/core/internals.py]: Sparse `astype(dtype, copy=False)` makes reinterpret cast, so I overrides `coerce_to_target_dtype` and set `copy=True` in `SparseBlock` class.
https://github.com/pandas-dev/pandas/pull/17386/files/2d94fbaccb5c2bd10a90b2ed77dcb3bade2b21c0#diff-e705e723b2d6e7c0e2a0443f80916abfR2690

Comment (Licht-T) [reply pandas/core/internals.py]: Sparse `astype(dtype, copy=False)` makes reinterpret cast, so I override `coerce_to_target_dtype` and set `copy=True` in `SparseBlock` class.
https://github.com/pandas-dev/pandas/pull/17386/files/2d94fbaccb5c2bd10a90b2ed77dcb3bade2b21c0#diff-e705e723b2d6e7c0e2a0443f80916abfR2690

File: pandas/core/internals.py (hunk 11 above)
Comment (Licht-T) [reply pandas/core/internals.py]: I added these codes because `coerce_to_target_dtype` does not work well in SparseBlock. This is only checking the type information of SparseArray and these procedure is also implemented in `IntBlock`, etc. I cannot figure out how to simplify. Any suggestion for simplifying?

**Comment by Licht-T** [comment]
@jreback Okay. I'll split the commit.

**Comment by Licht-T** [comment]
@jreback The huge commit is now split. If you have any questions, feel free to ask.

**Comment by Licht-T** [comment]
Rebased to change commit logs.

**Comment by Licht-T** [comment]
Rebased to change commit logs

File: pandas/core/internals.py [hunk 12]
```diff
@@ -613,7 +617,9 @@ def _can_hold_element(self, element):
             element = np.asarray(element)
             tipo = element.dtype.type
             return issubclass(tipo, dtype)
-        return isinstance(element, dtype)
+        else:
+            element_dtype = infer_dtype_from(element, pandas_dtype=True)[0]
+            return isinstance(element, dtype) or dtype == element_dtype
```
Comment (jreback) [inline pandas/core/internals.py]: huh?

File: pandas/core/internals.py [hunk 13]
```diff
@@ -995,7 +1001,7 @@ def coerce_to_target_dtype(self, other):
 
         if self.is_bool or is_object_dtype(dtype) or is_bool_dtype(dtype):
             # we don't upcast to bool
-            return self.astype(object)
+            return self.astype(object, copy=copy)
```
Comment (jreback) [inline pandas/core/internals.py]: I guess this is ok, though numpy ignores the ``copy=`` flag when dtype is ``object`` so no point in passing it if dtype is object

File: pandas/core/internals.py [hunk 14]
```diff
@@ -1026,18 +1032,18 @@ def coerce_to_target_dtype(self, other):
...
-                return self.astype(object)
+                return self.astype(object, copy=copy)
 
             raise AssertionError("possible recursion in "
                                  "coerce_to_target_dtype: {} {}".format(
                                      self, other))
 
         try:
-            return self.astype(dtype)
+            return self.astype(dtype, copy=copy)
```
Comment (jreback) [inline pandas/core/internals.py]: here i guess is ok

File: pandas/core/internals.py [hunk 15]
```diff
@@ -1382,6 +1388,11 @@ def where(self, other, cond, align=True, raise_on_error=True,
         if hasattr(other, 'reindex_axis'):
             other = other.values
 
+        if is_scalar(other) or is_list_like(other):
```
Comment (jreback) [inline pandas/core/internals.py]: huh?

File: pandas/core/internals.py [hunk 16]
```diff
@@ -1394,6 +1405,9 @@ def where(self, other, cond, align=True, raise_on_error=True,
         if not hasattr(cond, 'shape'):
             raise ValueError("where must have a condition that is ndarray "
                              "like")
+        else:
```
Comment (jreback) [inline pandas/core/internals.py]: huh? if you find yourself adding an if/then pretty much anywhere then you are doing it wrong. better to add the method to Sparse and call super; sometimes the super method may need a bit of refactor to make it more general.

File: pandas/core/internals.py [hunk 17]
```diff
@@ -1440,7 +1454,12 @@ def func(cond, values, other):
             if try_cast:
                 result = self._try_cast_result(result)
 
-            return self.make_block(result)
+            if isinstance(result, np.ndarray):
```
Comment (jreback) [inline pandas/core/internals.py]: again this is just completely confusing to do and makes the code way more complex. find a better way

File: pandas/core/internals.py [hunk 18]
```diff
@@ -1713,6 +1733,7 @@ class FloatBlock(FloatOrComplexBlock):
     is_float = True
     _downcast_dtype = 'int64'
 
+    @classmethod
```
Comment (jreback) [inline pandas/core/internals.py]: this is such a huge change, what is the purpose?

File: pandas/core/internals.py (hunk 11 above)
Comment (jreback) [reply pandas/core/internals.py]: I am not against this theorectially, but the implementation is fragile here. there are functions for validation already in the sparse classes.

File: pandas/core/internals.py [hunk 19]
```diff
@@ -2609,10 +2632,62 @@ class SparseBlock(NonConsolidatableMixIn, Block):
...
+            self.is_bool = True
+            self._can_hold_na = False
+        elif is_object_dtype(dtype):
+            self.is_object = True
+            self._can_hold_na = True
+        else:
+            self._can_hold_na = False
+
+    def _can_hold_element(self, element):
+        """ require the same dtype as ourselves """
```
Comment (jreback) [inline pandas/core/internals.py]: again this is so complex and adds so much techincal debt.

File: pandas/core/internals.py [hunk 20]
```diff
@@ -2769,9 +2858,15 @@ def sparse_reindex(self, new_index):
         return self.make_block_same_class(values, sparse_index=new_index,
                                           placement=self.mgr_locs)
 
+    def _try_coerce_result(self, result):
```
Comment (jreback) [inline pandas/core/internals.py]: so this is the right idea, though you shoul dgenerally have a function that intercepts a ndarray and creates a SparseArray; it should be called for most sparse methods.

**Review by jreback — CHANGES_REQUESTED** [review]
again you are doing too much in this PR. I require much simpler PR's that are built on top of each other. IOW pls break this apart. You *can* add tests for everything in a single PR (the first) and simply xfail tests that don't work. That will give you a basis to build on, and simplify understandbility of the reviews / PR / code. Adding paths via if/thens is not a good way forward here.

**Comment by jreback** [comment]
@Licht-T I believe you have all of the xfail tests in. pls rebase and update.

**Comment by Licht-T** [comment]
@jreback Yeah! Nothing stands in my way!

**Comment by Licht-T** [comment]
Now rebased.

**Comment by jreback** [comment]
@Licht-T can you see if you can simplify this, there seem to be lots of if/then cases. I think you might be able to define / override some sparse methods in internals to avoid this.

**Comment by jreback** [comment]
@Licht-T can you update?

**Comment by jreback** [comment]
closing as stale, though we'd definitely take a fixed up / rebased version updated for comments
